org.openhab.binding.zigbee/src/main/java/org/openhab/binding/zigbee/discovery/ZigBeeDiscoveryService.java

```diff
@@ -50,27 +57,61 @@
      */
     private final static int SEARCH_TIME = 60;
 
-    private final ZigBeeCoordinatorHandler coordinatorHandler;
-
     private DiscoveryServiceCallback discoveryServiceCallback;
 
-    public ZigBeeDiscoveryService(ZigBeeCoordinatorHandler coordinatorHandler) {
+    private final Set<ZigBeeCoordinatorHandler> coordinators = new CopyOnWriteArraySet<>();
+
+    private final Set<ZigBeeDiscoveryParticipant> participants = new CopyOnWriteArraySet<>();
+    private final Set<ThingTypeUID> supportedThingTypes = new CopyOnWriteArraySet<>();
+    private final Map<UID, ZigBeeNetworkNodeListener> registeredListeners = new ConcurrentHashMap<>();
+
+    public ZigBeeDiscoveryService() {
         super(SEARCH_TIME);
-        this.coordinatorHandler = coordinatorHandler;
-        logger.debug("Creating ZigBee discovery service for {}", coordinatorHandler.getThing().getUID());
+        logger.debug("Starting ZigBeeDiscoveryService");
     }
 
-    public void activate() {
-        logger.debug("Activating ZigBee discovery service for {}", coordinatorHandler.getThing().getUID());
+    @Reference(cardinality = ReferenceCardinality.MULTIPLE, policy = ReferencePolicy.DYNAMIC)
+    protected void addZigBeeDiscoveryParticipant(ZigBeeDiscoveryParticipant participant) {
+        participants.add(participant);
+        supportedThingTypes.addAll(participant.getSupportedThingTypeUIDs());
+    }
 
-        coordinatorHandler.addNetworkNodeListener(this);
+    protected void removeZigBeeDiscoveryParticipant(ZigBeeDiscoveryParticipant participant) {
+        supportedThingTypes.removeAll(participant.getSupportedThingTypeUIDs());
```

Comment:
This might be an edge case, but consider a thing type that is supported by two discovery participants. In this case, it is not ok to remove all supported thing types when removing one of the two discovery participants.

One remedy here could be to not store the supported thing types in their own Set, but rather to construct the set each time the `getSupportedThingTypes` method is called.

Reply:
I guess we could also rebuild the full set whenever a participant is removed - that way it will always be correct.

I took this from the Bluetooth code that Kai wrote, so I assumed that this was ok.

Reply:
Ah, ok. It might be an edge case. @kaikreuzer Do you recall why it was implemented like this for the bluetooth binding?